Source/Particles/WarpXParticleContainer.cpp

```diff
@@ -1420,27 +1420,26 @@ std::array<ParticleReal, 3> WarpXParticleContainer::meanParticleVelocity(bool lo
     return mean_v;
 }
 
-amrex::ParticleReal WarpXParticleContainer::maxParticleVelocity(bool local) {
+amrex::ParticleReal WarpXParticleContainer::maxParticleVelocity(bool local) const {
 
-    amrex::ParticleReal max_v = 0.0;
+    const amrex::ParticleReal inv_clight_sq = 1.0_prt/(PhysConst::c*PhysConst::c);
 
-    const int nLevels = finestLevel();
-    for (int lev = 0; lev <= nLevels; ++lev)
-    {
-
-#ifdef AMREX_USE_OMP
-#pragma omp parallel reduction(max:max_v)
-#endif
-        for (WarpXParIter pti(*this, lev); pti.isValid(); ++pti)
-        {
-            auto& ux = pti.GetAttribs(PIdx::ux);
-            auto& uy = pti.GetAttribs(PIdx::uy);
-            auto& uz = pti.GetAttribs(PIdx::uz);
-            for (unsigned long i = 0; i < ux.size(); i++) {
-                max_v = std::max(max_v, std::sqrt(ux[i]*ux[i] + uy[i]*uy[i] + uz[i]*uz[i]));
-            }
-        }
-    }
+    using PType = typename WarpXParticleContainer::SuperParticleType;
+    amrex::ReduceOps<amrex::ReduceOpMax> reduce_ops;
+    const auto reduce_res = amrex::ParticleReduce<amrex::ReduceData<amrex::ParticleReal>>(
+            *this,
+            [=] AMREX_GPU_HOST_DEVICE(const PType &p) noexcept -> amrex::GpuTuple<amrex::ParticleReal>
+            {
+                const auto ux = p.rdata(PIdx::ux);
+                const auto uy = p.rdata(PIdx::uy);
+                const auto uz = p.rdata(PIdx::uz);
+                return {(ux*ux + uy*uy + uz*uz) * inv_clight_sq,};
```

Comment:
NVCC issue: We have to double check if you want to use a different syntax for the return line or skip the training return type to make it happy.

Maybe try to use `amrex::makeTuple(...)`?

Reply:
@RemiLehe also spotted an overload of this without a tuple (for scalars) - will post